lib/puppet/etc.rb

```diff
@@ -0,0 +1,154 @@
+require 'puppet/util/character_encoding'
+# Wrapper around Ruby Etc module allowing us to manage encoding in a single
+# place.
+# This represents a subset of Ruby's Etc module, only the methods required by Puppet.
+# Etc returns strings in variable encoding depending on
+# environment. For Puppet we specifically want UTF-8 as our input from the Etc
+# module - which is our input for many resource instance 'is' values. The
+# associated 'should' value will basically always be coming from Puppet in
+# UTF-8. Etc is defined for Windows but calls to it return nil.
+#
+# We only use Etc for retrieving existing property values from the system. For
+# setting property values, providers leverage system tools (i.e., `useradd`)
+#
+# @api private
+module Puppet::Etc
+  class << self
+    # Etc::getgrent returns an Etc::Group struct object
+    # On first call opens /etc/group and returns parse of first entry. Each subsquent call
+    # returns new struct the next entry or nil if EOF. Call ::endgrent to close file.
+    def getgrent
+      group_entry = ::Etc.getgrent
+      if !group_entry.nil?
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
```

Comment:
we could simply all of the `get*` methods by making the `convert_*` method nil safe:

```ruby
def getgrent
  to_utf8 ::Etc.getgrent)
end

def to_utf8(entry)
  return nil if entry.nil?
  ...
end
```